staging/src/k8s.io/apimachinery/pkg/api/meta/help_test.go

```diff
@@ -0,0 +1,870 @@
+				},
+				Spec: FooSpec{
+					Flied: i,
+				},
+			},
+		}
+	}
+	return out
+}
+
+func fakeUnstructuredList(numItems int) runtime.Unstructured {
+	out := &unstructured.UnstructuredList{
+		Items: make([]unstructured.Unstructured, numItems),
+	}
+
+	for i := range out.Items {
+		out.Items[i] = unstructured.Unstructured{
+			Object: map[string]interface{}{
+				"apiVersion": "v1",
+				"kind":       "Pod",
+				"metadata": map[string]interface{}{
+					"creationTimestamp": nil,
+					"name":              strconv.Itoa(i),
+				},
+				"spec": map[string]interface{}{
+					"hostname": "example.com",
+				},
+				"status": map[string]interface{}{},
+			},
+		}
+	}
+	return out
+}
+
+func fakeFooList(numItems int) *FooList {
+	out := &FooList{
+		Items: make([]Foo, numItems),
+	}
+
+	for i := range out.Items {
+		out.Items[i] = Foo{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: "sample.org/v1",
+				Kind:       "Foo",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      strconv.Itoa(i),
+				Namespace: "default",
+				Labels: map[string]string{
+					"label-key-1": "label-value-1",
+				},
+				Annotations: map[string]string{
+					"annotations-key-1": "annotations-value-1",
+				},
+			},
+			Spec: FooSpec{
+				Flied: i,
+			},
+		}
+	}
+	return out
+}
+
+func TestMemoryBehaviorExtractList(t *testing.T) {
+	tests := []struct {
+		name            string
+		generateFunc    func(num int) (list runtime.Object, firstObjectPointer interface{})
+		numObject       int
+		shouldBeCleared bool
+	}{
+		{
+			name: "FooList",
+			generateFunc: func(num int) (list runtime.Object, firstObjectPointer interface{}) {
+				out := fakeFooList(num)
+				return out, &out.Items[0]
+			},
+			numObject:       fakeObjectItemsNum,
+			shouldBeCleared: true,
+		},
+		{
+			name: "SampleList",
+			generateFunc: func(num int) (list runtime.Object, firstObjectPointer interface{}) {
+				out := fakeSampleList(num)
+				return out, &out.Items[0]
+			},
+			numObject:       fakeObjectItemsNum,
+			shouldBeCleared: false,
+		},
+		{
+			name: "RawExtensionList",
+			generateFunc: func(num int) (list runtime.Object, firstObjectPointer interface{}) {
+				out := fakeExtensionList(num)
+				return out, &out.Items[0]
+			},
+			numObject:       fakeObjectItemsNum,
+			shouldBeCleared: true,
+			// NOTE:
+			// although EachListItem takes the address of the elements of items,
+			// but it uses items.Object instead of items.
+			// so the items slice will be cleared.
+		},
+		{
+			name: "UnstructuredList",
+			generateFunc: func(num int) (list runtime.Object, firstObjectPointer interface{}) {
+				out := fakeUnstructuredList(num)
+				unstructuredList := out.(*unstructured.UnstructuredList)
+				return out, &unstructuredList.Items[0]
+			},
+			numObject:       fakeObjectItemsNum,
+			shouldBeCleared: false,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			ctx, cancel := context.WithTimeout(context.Background(), wait.ForeverTestTimeout)
```

Comment:
I can’t, reasons:

- Because the goal of this optimization is to be able to prove that unnecessary memory is indeed GC
- ExtractListWithAlloc if objList is of type unstructured.UnstructuredList. Because of the use of shallow copy, modifying the APIVersion will still cause the original Object elements in objList.Items[0] to be modified.